File: Python/basicget.py
"""Example of getting a message from an IBM MQ queue
"""

# -*- coding: utf-8 -*-
# Copyright 2019, 2025 IBM
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging

import ibmmq as mq
from utils.env import EnvStore

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

WAIT_INTERVAL = 5  # Seconds

def connect(index):
    """ Establish connection to MQ Queue Manager """
    logger.info('Establishing Connection with MQ Server')

    # Set credentials if they have been provided
    csp = mq.CSP()
    csp.CSPUserId = EnvStore.getenv_value(EnvStore.APP_USER, index)
    csp.CSPPassword = EnvStore.getenv_value(EnvStore.APP_PASSWORD, index)
    if csp.CSPUserId is None:
        csp = None

    try:
        cd = None
        if not EnvStore.is_ccdt_available():
            logger.info('CCDT URL export is not set, will be using json environment client connections settings')

            cd = mq.CD(Version=mq.CMQXC.MQCD_VERSION_11)
            cd.ChannelName = MQDetails[EnvStore.CHANNEL]
            cd.ConnectionName = conn_info
            cd.ChannelType = mq.CMQXC.MQCHT_CLNTCONN
            cd.TransportType = mq.CMQXC.MQXPT_TCP

            logger.info('Checking cipher details')
            # If a cipher is set then set the TLS settings
            if MQDetails[EnvStore.CIPHER]:
                logger.info('Making use of cipher details')
                cd.SSLCipherSpec = MQDetails[EnvStore.CIPHER]

        # Key repository is not specified in CCDT so look in environment settings
        # Create an empty SCO object
        sco = mq.SCO()
        if MQDetails[EnvStore.KEY_REPOSITORY]:
            logger.info('Setting Key repository')
            sco.KeyRepository = MQDetails[EnvStore.KEY_REPOSITORY]

        qmgr = mq.QueueManager(None)
        qmgr.connect_with_options(MQDetails[EnvStore.QMGR],
                                  csp=csp,
                                  cd=cd, sco=sco)
        return qmgr
    except mq.MQMIError as e:
        logger.error('Error connecting: %s', e)
        return None

def get_queue():
    """Establish access to a Queue"""
    logger.info('Opening a Queue')
    try:
        q = mq.Queue(qmgr)

        od = mq.OD()
        od.ObjectName = MQDetails[EnvStore.QUEUE_NAME]

        open_options = mq.CMQC.MQOO_INPUT_AS_Q_DEF
        q.open(od, open_options)

        return q
    except mq.MQMIError as e:
        logger.error('Error opening queue: %s', e)
        return None

def get_messages():
    """Get messages from the queue"""
    logger.info('Attempting gets from Queue')

    # Message Descriptor
    md = mq.MD()

    # Get Message Options
    # MQGMO_NO_PROPERTIES indicates that JMS headers are to be stripped
    # off the message during the get. This can also be done by calling
    # get_no_jms() on the queue instead of get().
    gmo = mq.GMO()
    gmo.Options = (mq.CMQC.MQGMO_WAIT |
                   mq.CMQC.MQGMO_FAIL_IF_QUIESCING |
                   mq.CMQC.MQGMO_NO_PROPERTIES)

    gmo.WaitInterval = WAIT_INTERVAL * 1000  # Convert to milliseconds

    keep_running = True
    while keep_running:
        try:
            # Reset the MsgId, CorrelId & GroupId so that we can reuse
            # the same 'md' object again.
            md.MsgId = mq.CMQC.MQMI_NONE
            md.CorrelId = mq.CMQC.MQCI_NONE
            md.GroupId = mq.CMQC.MQGI_NONE

            # Wait up to to gmo.WaitInterval for a new message.
            # message = queue.get_no_jms(None, md, gmo)
            message = queue.get(None, md, gmo)

            # Process the message here..
            logger.info('Have message from input queue: %s', message.decode())

        except mq.MQMIError as e:
            if e.comp == mq.CMQC.MQCC_FAILED and e.reason == mq.CMQC.MQRC_NO_MSG_AVAILABLE:
                # No messages, we should more on to next connection endpoint if there is one.
                logger.info('No more messages found on this connection')
                keep_running = False
            else:
                # Some other error condition.
                raise

        except (UnicodeDecodeError, ValueError) as e:
            logger.info('Message is not valid json')
            logger.info(e)
            logger.info(message)
            continue

        except KeyboardInterrupt:
            logger.info('Have received a keyboard interrupt')
            keep_running = False

def build_mq_details(index):
    """Create the connection details for the queue manager"""
    for key in [EnvStore.QMGR, EnvStore.QUEUE_NAME, EnvStore.CHANNEL, EnvStore.HOST,
                EnvStore.PORT, EnvStore.KEY_REPOSITORY, EnvStore.CIPHER]:
        MQDetails[key] = EnvStore.getenv_value(key, index)

# Application logic starts here
logger.info('Application "BasicGet" is starting')

envStore = EnvStore()
envStore.set_env()

MQDetails = {}

qmgr = None
queue = None

numEndPoints = envStore.get_endpoint_count()
logger.info('There are %d connections', numEndPoints)

# Loop through the connection options. If one succeeds, do the
# work and then quit.
for index, conn_info in envStore.get_next_connection_string():
    logger.info('Using Connection String %s', conn_info)

    build_mq_details(index)

    qmgr = connect(index)
    if qmgr is not None:
        queue = get_queue()
        if queue is not None:
            get_messages()
            queue.close()

        qmgr.disconnect()
        break

MQDetails.clear()

logger.info('Application is ending')

File: Go/src/basicget.go
/**
 * Copyright 2019, 2026 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package main

import (
	"log"
	"mqdevpatterns/src/mqsamputils"
	"os"
	"strings"

	"github.com/ibm-messaging/mq-golang/v5/ibmmq"
)

var logger = log.New(os.Stdout, "MQ Get: ", log.LstdFlags)

type message struct {
	Greeting string `json:"greeting"`
	Value    int    `json:"value"`
}

// Main entry to Get application
// Creates connection to Queue Manager
func main() {

	logger.Println("Application is starting")

	mqsamputils.EnvSettings.LogSettings()

	qMgr, err := mqsamputils.CreateConnection(mqsamputils.FULL_STRING)
	if err != nil {
		logger.Fatalln("Unable to establish connection to server")
	}
	defer qMgr.Disc()

	qObject, err := openQueue(qMgr)
	if err != nil {
		logger.Fatalln("Unable to open queue")
	}
	defer qObject.Close(0)

	getMessage(qObject)

	logger.Println("Application is ending")
}

func openQueue(qMgr ibmmq.MQQueueManager) (ibmmq.MQObject, error) {
	// Create the Object Descriptor that allows us to give the queue name
	mqod := ibmmq.NewMQOD()
	mqod.ObjectType = ibmmq.MQOT_Q
	mqod.ObjectName = mqsamputils.EnvSettings.QueueName

	openOptions := ibmmq.MQOO_INPUT_EXCLUSIVE

	logger.Printf("Attempting to open queue %s", mqod.ObjectName)
	qObject, err := qMgr.Open(mqod, openOptions)

	if err != nil {
		logError(err)
	} else {
		logger.Printf("Opened object %s", qObject.Name)
	}

	return qObject, err
}

func logError(err error) {
	logger.Println(err)
	os.Exit(1)
}

func getMessage(qObject ibmmq.MQObject) {
	logger.Println("Getting message from queue")
	var err error
	msgAvail := true

	for msgAvail == true && err == nil {
		var datalen int

		// The PUT requires control structures, the Message Descriptor (MQMD)
		// and Put Options (MQPMO). Create those with default values.
		getmqmd := ibmmq.NewMQMD()
		gmo := ibmmq.NewMQGMO()

		// The default options are OK, but it's always
		// a good idea to be explicit about transactional boundaries as
		// not all platforms behave the same way.
		gmo.Options = ibmmq.MQGMO_NO_SYNCPOINT

		// Set options to wait for a maximum of 3 seconds for any new message to arrive
		gmo.Options |= ibmmq.MQGMO_WAIT
		gmo.WaitInterval = 3 * 1000 // The WaitInterval is in milliseconds

		// Create a buffer for the message data. This one is large enough
		// for the messages put by the amqsput sample.
		buffer := make([]byte, 1024)

		// Now try to get the message
		datalen, err = qObject.Get(getmqmd, gmo, buffer)

		if err != nil {
			msgAvail = false

			mqret := err.(*ibmmq.MQReturn)
			if mqret.MQRC == ibmmq.MQRC_NO_MSG_AVAILABLE {
				// If there's no message available, then don't treat that as a real error as
				// it's an expected situation. But we still use it as an opportunity to exit.
				err = nil
				logger.Println("No more messages on this endpoint")
			} else {
				logError(err)
			}
		} else {
			// Assume the message is a printable string
			logger.Printf("Got message of length %d: ", datalen)
			logger.Println("  " + strings.TrimSpace(string(buffer[:datalen])))
		}
	}
}
